p2p/protocol/identify/peer_loop.go

```diff
@@ -0,0 +1,209 @@
+
+		case fnc := <-ph.evalTestCh:
+			fnc()
+
+		case <-ph.ctx.Done():
+			return
+		}
+	}
+}
+
+func (ph *peerHandler) applyDelta(mes *pb.Delta) {
+	for _, p1 := range mes.RmProtocols {
+		for j, p2 := range ph.idMsgSnapshot.Protocols {
+			if p2 == p1 {
+				ph.idMsgSnapshot.Protocols[j] = ph.idMsgSnapshot.Protocols[len(ph.idMsgSnapshot.Protocols)-1]
+				ph.idMsgSnapshot.Protocols = ph.idMsgSnapshot.Protocols[:len(ph.idMsgSnapshot.Protocols)-1]
+			}
+		}
+	}
+
+	for _, p := range mes.AddedProtocols {
+		ph.idMsgSnapshot.Protocols = append(ph.idMsgSnapshot.Protocols, p)
+	}
+}
+
+func (ph *peerHandler) openStream(protos []string) (network.Stream, error) {
+	// wait for the other peer to send us an Identify response on "all" connections we have with it
+	// so we can look at it's supported protocols and avoid a multistream-select roundtrip to negotiate the protocol
+	// if we know for a fact that it dosen't support the protocol.
+	conns := ph.ids.Host.Network().ConnsToPeer(ph.pid)
+	for _, c := range conns {
+		select {
+		case <-ph.ids.IdentifyWait(c):
+		case <-ph.ctx.Done():
+			return nil, ph.ctx.Err()
+		}
+	}
+	pstore := ph.ids.Host.Peerstore()
+	if sup, err := pstore.SupportsProtocols(ph.pid, protos...); err != nil && len(sup) == 0 {
+		return nil, errProtocolNotSupported
+	}
+
+	// negotiate a stream without opening a new connection as we "should" already have a connection.
+	ctx, cancel := context.WithTimeout(ph.ctx, 30*time.Second)
+	defer cancel()
+	ctx = network.WithNoDial(ctx, "should already have connection")
+
+	// newstream will open a stream on the first protocol the remote peer supports from the among
+	// the list of protocols passed to it.
+	s, err := ph.ids.Host.NewStream(ctx, ph.pid, protocol.ConvertFromStrings(protos)...)
+	if err != nil {
+		log.Warnf("failed to open %s stream with peer %s, err=%s", protos, ph.pid.Pretty(), err)
+		return nil, err
+	}
+
+	return s, err
+}
+
+func (ph *peerHandler) mkDelta() *pb.Delta {
+	old := ph.idMsgSnapshot.GetProtocols()
+	curr := ph.ids.Host.Mux().Protocols()
+
+	oldProtos := make(map[string]struct{}, len(old))
+	currProtos := make(map[string]struct{}, len(curr))
+
+	for i := range old {
+		oldProtos[old[i]] = struct{}{}
+	}
+
+	for i := range curr {
+		currProtos[curr[i]] = struct{}{}
+	}
+
+	var added []string
+	var removed []string
+
+	// has it been added ?
+	for p := range currProtos {
+		if _, ok := oldProtos[p]; !ok {
+			added = append(added, p)
+		}
+	}
+
+	// has it been removed ?
+	for p := range oldProtos {
+		if _, ok := currProtos[p]; !ok {
+			removed = append(removed, p)
+		}
+	}
+
+	return &pb.Delta{
+		AddedProtocols: added,
+		RmProtocols:    removed,
+	}
+}
+
+func (ph *peerHandler) sendMessage(s network.Stream, mes *pb.Identify) error {
+	defer helpers.FullClose(s)
+	c := s.Conn()
+	if err := ggio.NewDelimitedWriter(s).WriteMsg(mes); err != nil {
+		log.Warnf("error while sending %s update to %s: err=%s", s.Protocol(), c.RemotePeer(), err)
```

Comment:
nit: prefer `log.Warnw("error while sending identify update", "protocol", s.Protocol(), "peer", c.RemotePeer(), "error", err)`. This gives us nice structured logs we can index and query.

(we _used_ to use Warnf and friends but we're switching to a new logging backend)

Reply:
Doing this everywhere now.